lib/runtime/storage/trie.go

```diff
@@ -17,97 +17,129 @@ import (
 // TrieState is a wrapper around a transient trie that is used during the course of executing some runtime call.
 // If the execution of the call is successful, the trie will be saved in the StorageState.
 type TrieState struct {
-	t       *trie.Trie
-	oldTrie *trie.Trie // this is the trie before BeginStorageTransaction is called. set to nil if it isn't called
-	lock    sync.RWMutex
+	mtx          sync.RWMutex
+	transactions []*trie.Trie
```

Comment:
If we change the `transactions []*trie.Trie` to `transactions *list.List` the implementation could be simpler since we can use that list as a `stack` of transactions